README.md

```diff
@@ -1110,6 +1110,175 @@ Generally, either the default implementation or the override closures should pro
 
 > It is important to keep in mind that the `subdelegates` are initialized and destroyed in the default implementation. Be careful when subclassing to not introduce memory leaks.
 
+### Adapting and Retrying Requests
+
+Most web services these days are behind some sort of authentication system. One of the more common ones today is OAuth. This generally involves generating an access token authorizing your application or user to call the various supported web services. While creating these initial access tokens can be laborsome, it can be even more complicated when your access token expires and you need to fetch a new one. There are many thread-safety issues that need to be considered.
+
+The `RequestAdapter` and `RequestRetrier` protocols were created to make it much easier to create a thread-safe authentication system for a specific set of web services.
+
+#### RequestAdapter
+
+The `RequestAdapter` protocol allows each `Request` made on a `SessionManager` to be inspected and adapted before being created. One very specific way to use an adapter is to append an `Authorization` header to requests behind a certain type of authentication.
+
+```swift
+class AccessTokenAdapter: RequestAdapter {
+	private let accessToken: String
+	
+	init(accessToken: String) {
+		self.accessToken = accessToken
+	}
+
+	func adapt(_ urlRequest: URLRequest) -> URLRequest {
+	    var urlRequest = urlRequest
+
+	    if urlRequest.urlString.hasPrefix("https://httpbin.org") {
+		    urlRequest.setValue("Bearer " + accessToken, forHTTPHeaderField: "Authorization")
+	    }
+
+	    return urlRequest
+	}
+}
+
+let sessionManager = SessionManager()
+sessionManager.adapter = AccessTokenAdapter(accessToken: "1234")
+
+sessionManager.request("https://httpbin.org/get", withMethod: .get)
+```
+
+#### RequestRetrier
+
+The `RequestRetrier` protocol allows a `Request` that encountered an `Error` while being executed to be retried. When using both the `RequestAdapter` and `RequestRetrier` protocols together, you can create credential refresh systems for OAuth1, OAuth2, Basic Auth and even exponential backoff retry policies. The possibilities are endless. Here's a short example of how you could implement a refresh flow for OAuth2 access tokens.
```

Comment:
Seems like there should be a simpler example or walkthrough of the `RequestRetrier` protocol before jumping directly into an OAuth2 implementation. From looking at it I can't tell what's actually required for the protocol here and what's part of the OAuth2 flow.